pkg/backends/calico/routes.go

```diff
@@ -195,32 +196,49 @@ func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
 	rg.Lock()
 	if rg.advertiseThisService(svc, ep) {
 		rg.svcRouteMap[key] = []string{svc.Spec.ClusterIP + "/32"}
-	} else if _, exists := rg.svcRouteMap[key]; exists {
+		rg.client.AddStaticRoutes([]string{svc.Spec.ClusterIP + "/32"})
+	} else if routes, exists := rg.svcRouteMap[key]; exists {
+		rg.client.DeleteStaticRoutes(routes)
 		delete(rg.svcRouteMap, key)
 	}
 	rg.Unlock()
-
-	rg.updateRoutes()
 }
 
 // advertiseThisService returns true if this service should be advertised on this node,
 // false otherwise.
 func (rg *routeGenerator) advertiseThisService(svc *v1.Service, ep *v1.Endpoints) bool {
-	// always set if externalTrafficPolicy != local
+	logc := log.WithField("svc", fmt.Sprintf("%s/%s", svc.Namespace, svc.Name))
+
+	// do nothing if the svc is not a relevant type
+	if (svc.Spec.Type != v1.ServiceTypeClusterIP) && (svc.Spec.Type != v1.ServiceTypeNodePort) && (svc.Spec.Type != v1.ServiceTypeLoadBalancer) {
+		logc.Debugf("Skipping service with cluster type %s", svc.Spec.Type)
+		return false
+	}
+
+	// also do nothing if the clusterIP is empty or None
+	if svc.Spec.ClusterIP == "" || svc.Spec.ClusterIP == "None" {
+		logc.Debugf("Skipping service with no cluster IP %s", svc.Spec.Type)
+		return false
+	}
+
+	// we only need to advertise local services, since we advertise the entire cluster IP range.
 	if svc.Spec.ExternalTrafficPolicy != v1.ServiceExternalTrafficPolicyTypeLocal {
-		return true
+		logc.Debugf("Skipping service with non-local external traffic policy '%s'", svc.Spec.ExternalTrafficPolicy)
+		return false
 	}
 
-	// add svc clusterIP if node contains at least one endpoint for svc
+	// advertise clusterIP if node contains at least one endpoint for svc
 	for _, subset := range ep.Subsets {
 		// not interested in subset.NotReadyAddresses
 		for _, address := range subset.Addresses {
 			if address.NodeName == nil || *address.NodeName != rg.nodeName {
 				continue
 			}
+			logc.Debugf("Advertising local service")
 			return true
```

Comment:
It's a small point, but I think this would read more naturally:
```
			if address.NodeName != nil && *address.NodeName == rg.nodeName {
 			    logc.Debugf("Advertising local service")
 			    return true
 			}
```